clang/lib/Driver/ToolChains/HLSL.cpp

```diff
@@ -122,6 +126,30 @@ std::string tryParseProfile(StringRef Profile) {
     return "";
 }
 
+bool isLegalValidatorVersion(StringRef ValVersionStr, std::string &ErrorMsg) {
```

Comment:
This pattern of passing in an error string by reference then feeding that to Diag, is not really idiomatic. Instead you should emit the diagnostics where you're currently setting the ErrorMsg string, and probably have different diagnostics for each thing that can be wrong.